[READY] Implement shutdown handler

micbou opened the pull request:
This PR depends on https://github.com/Valloric/ycmd/pull/282.

We don't want to block Vim exit if the server takes too much time to answer the shutdown request so we use a timeout of 100ms for this request. In my experience, it takes ~5ms.

The `YcmStopServer` command only exists for testing purpose. It will be removed in the final version of this PR.

Fixes #876.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2245)
<!-- Reviewable:end -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice!

---

Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 130 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMjlDgdNTQ8e8ufKf1V:-KMjlDgdNTQ8e8ufKf1W:-1013055074) ([raw file](https://github.com/valloric/youcompleteme/blob/48d1557e199899a0759d80ed8b36958e0fe6958c/python/ycm/youcompleteme.py#L130)):*
> ```Python
>   def _SetupServer( self ):
>     self._available_completers = {}
>     self._user_notified_about_crash = False
> ```

Actually, this should be inside `__init__`. The idea is to initialize _all_ member var inside `__init__` if for any reason, then so that future readers have an easy way of knowing what are all the member variables used in this class.

Of course, there's nothing wrong with setting a different value to the member var in other functions, but they should be declared in `__init__`. (I use "declare" here loosely because there's no such concept in Python; values are merely assigned. Here by "declare" I mean "assigned for the first time.")

Any instances where we have member vars declared outside of `__init__` are an oversight. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KMjmP7KXQDDehxBVwxe:1300102931)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 130 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMjlDgdNTQ8e8ufKf1V:-KMjrRVZ36wHfrEFJqyy:276585145) ([raw file](https://github.com/valloric/youcompleteme/blob/48d1557e199899a0759d80ed8b36958e0fe6958c/python/ycm/youcompleteme.py#L130)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Actually, this should be inside `__init__`. The idea is to initialize _all_ member var inside `__init__` if for any reason, then so that future readers have an easy way of knowing what are all the member variables used in this class.
> 
> Of course, there's nothing wrong with setting a different value to the member var in other functions, but they should be declared in `__init__`. (I use "declare" here loosely because there's no such concept in Python; values are merely assigned. Here by "declare" I mean "assigned for the first time.")
> 
> Any instances where we have member vars declared outside of `__init__` are an oversight. 
</details>
I see but these variables need to be reset when restarting the server. If we initialize them inside `__init__`, we will set them twice to the same value when starting the server for the first time. Should we reset them in the `RestartServer` method instead? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 130 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMjlDgdNTQ8e8ufKf1V:-KMk1ciGIsoJDsjZ7D3j:-232857517) ([raw file](https://github.com/valloric/youcompleteme/blob/48d1557e199899a0759d80ed8b36958e0fe6958c/python/ycm/youcompleteme.py#L130)):*
> we will set them twice to the same value when starting the server for the first time. 

Why is this a problem? 

I'm fine with setting the var to `None` in the ctor and then to `False` elsewhere if that helps. But setting it to `False` straight away still seems like the best choice.

> Should we reset them in the RestartServer method instead?

Up to you. All I'd like to see is all member vars set to _something_ in the ctor so that the reader can easily see what are all the member vars used in the class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Awesome.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one comment.

---

Reviewed 7 of 9 files at r1, 3 of 4 files at r2.
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 779 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMoOBUvpz7Q7zq_vDnI:-KMoOBUvpz7Q7zq_vDnJ:247078527) ([raw file](https://github.com/valloric/youcompleteme/blob/44d12e80747d1eddb2a9a2294b6aed2d26896e6a/autoload/youcompleteme.vim#L779)):*
> ```VimL
> 
> function! s:StopServer()
>   py ycm_state.StopServer()
> ```

Does this work on py3 ? Isn't the command :py3 ?

IIRC we have a method or some script-local variable used to call the correct `py` command for PY2/PY3 ? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KMoPfC2zPzXNIxjseUu:-1007301825)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 1 files at r3.
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 779 \[r2\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMoOBUvpz7Q7zq_vDnI:-KMoQFB_nrKkdi1_A4WQ:-301847483) ([raw file](https://github.com/valloric/youcompleteme/blob/44d12e80747d1eddb2a9a2294b6aed2d26896e6a/autoload/youcompleteme.vim#L779)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary>
> Does this work on py3 ? Isn't the command :py3 ?
> 
> IIRC we have a method or some script-local variable used to call the correct `py` command for PY2/PY3 ? 
</details>
Oops, I added the command a long time ago before we bring Python 3 support. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 1 of 4 files at r2.
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 130 \[r1\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMjlDgdNTQ8e8ufKf1V:-KMod5byRHWaTGfMzPIk:-2014790819) ([raw file](https://github.com/valloric/youcompleteme/blob/48d1557e199899a0759d80ed8b36958e0fe6958c/python/ycm/youcompleteme.py#L130)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> > we will set them twice to the same value when starting the server for the first time. 
> 
> Why is this a problem? 
> 
> I'm fine with setting the var to `None` in the ctor and then to `False` elsewhere if that helps. But setting it to `False` straight away still seems like the best choice.
> 
> > Should we reset them in the RestartServer method instead?
> 
> Up to you. All I'd like to see is all member vars set to _something_ in the ctor so that the reader can easily see what are all the member vars used in the class.
</details>
These two variables are now initialized in the constructor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 6 of 9 files at r1, 4 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[.gitmodules, line 6 \[r3\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMpVM8919EE5FKNbP_7:-KMpVM8919EE5FKNbP_8:-304380588) ([raw file](https://github.com/valloric/youcompleteme/blob/7027ff89903e06bbf80f196c083a488a0b06dda1/.gitmodules#L6)):*
> ```
> [submodule "third_party/ycmd"]
> 	path = third_party/ycmd
> 	url = https://github.com/micbou/ycmd
> ```

Don't forget to remove this. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Looks good, though it looks like there's a problem with the CI... do you need a rebase, or is it flaking?

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KMpvClgjeYxa49AckxU:-843681768)*
<!-- Sent from Reviewable.io -->

micbou commented:
Yes, a rebase was needed.

---

Reviewed 5 of 9 files at r1, 3 of 4 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KMq1CBweJ7ei59N0l_x:-788381189)*
<!-- Sent from Reviewable.io -->

micbou commented:
Updated to latest ycmd. This is ready.

---

Reviewed 3 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[.gitmodules, line 6 \[r3\]](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-KMpVM8919EE5FKNbP_7:-KNM4B21ci0o8b4Hn_xE:1740281843) ([raw file](https://github.com/valloric/youcompleteme/blob/7027ff89903e06bbf80f196c083a488a0b06dda1/.gitmodules#L6)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> Don't forget to remove this. :)
</details>
Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KNM4EB_4n3S4_HCKLG5:1208409680)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4, 3 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KNMZfc9U79bAgC0BzZt:1417088721)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> @homu r=puremourning

---

Reviewed 5 of 9 files at r1, 1 of 4 files at r2, 1 of 1 files at r4, 3 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2245#-:-KNMtus3HuuPwAoaCWEI:327595638)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 4e94d7c has been approved by `puremourning`

<!-- @homu r=puremourning 4e94d7cad69c05d280a3cf93c45df51e10f47b20 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/146830861)
